rp2040-hal/src/watchdog.rs

```diff
@@ -1,3 +1,91 @@
 //! Watchdog
 // See [Chapter 4 Section 7](https://datasheets.raspberrypi.org/rp2040/rp2040_datasheet.pdf) for more details
-// TODO
+use crate::pac::WATCHDOG;
+use embedded_hal::watchdog;
+use embedded_time::{duration, fixed_point::FixedPoint};
+
+/// Watchdog peripheral
+pub struct Watchdog {
+    watchdog: WATCHDOG,
+    delay_ms: u32,
+}
+
+impl Watchdog {
+    /// Create a new [`Watchdog`]
+    pub fn new(watchdog: WATCHDOG) -> Self {
+        Self {
+            watchdog,
+            delay_ms: 0,
+        }
+    }
+
+    /// Starts tick generation on clk_tick which is driven from clk_ref.
+    ///
+    /// # Arguments
+    ///
+    /// * `cycles` - Total number of tick cycles before the next tick is generated.
+    pub fn enable_tick_generation(&mut self, cycles: u8) {
+        const WATCHDOG_TICK_ENABLE_BITS: u32 = 0x200;
+
+        self.watchdog
+            .tick
+            .write(|w| unsafe { w.bits(WATCHDOG_TICK_ENABLE_BITS | cycles as u32) })
+    }
+
+    /// Defines whether or not the watchdog timer should be paused when processor(s) are in debug mode
+    /// or when JTAG is accessing bus fabric
+    ///
+    /// # Arguments
+    ///
+    /// * `pause` - If true, watchdog timer will be paused
+    pub fn pause_on_debug(&mut self, pause: bool) {
+        self.watchdog.ctrl.write(|w| {
+            w.pause_dbg0()
+                .bit(pause)
+                .pause_dbg1()
+                .bit(pause)
+                .pause_jtag()
+                .bit(pause)
+        })
+    }
+
+    fn load_counter(&self, counter: u32) {
```

Comment:
Isn't this unsafe? What is they give a value greater than MAX_PERIOD here?

Maybe the period should be a newtype with a const new function that panics if out of bounds, so it can blow up at compile time in a const context.

Reply:
Thanks for the feedback!

> Isn't this unsafe? What is they give a value greater than MAX_PERIOD here?
I'm not sure I understand what you mean by `unsafe`.  If the concern is overwriting reserved bits, the value is masked to 0xFF_FFFF [here](https://github.com/rp-rs/rp2040-pac/blob/4006f1937b8f57c4b4d6543bfd7aa7158a640fb0/src/watchdog/load.rs#L19).  Otherwise, the function is private and used exclusively by the Watchdog HAL trait implementations below (which panics if the provided period exceeds `MAX_PERIOD`)

> Maybe the period should be a newtype with a const new function that panics if out of bounds, so it can blow up at compile time in a const context.

I'm not sure I know this pattern, could you provide an example?

Reply:
Ah, OK. I hadn't spotted that was was a non-pub function and yes, I can see that the PAC limits the values to MAX_PERIOD anyway.